File: python/309_Best Time to Buy and Sell Stock with Cooldown.py
#!/usr/bin/env python2

class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        if not prices or len(prices) < 2:
            return 0
        buy = [0] * len(prices)
        sell = [0] * len(prices)

        buy[0] = -prices[0]
        buy[1] = max(-prices[0], -prices[1])
        sell[0] = 0
        sell[1] = max(sell[0], buy[0]+prices[1])
        for i in range(2, len(prices)):
            buy[i]= max(buy[i-1], sell[i-2]-prices[i])
            sell[i] = max(sell[i-1], buy[i-1]+prices[i])

        return sell[-1]

File: java/src/brand_new/_309_Best_Time_to_Buy_and_Sell_Stock_with_Cooldown.java
package brand_new;

public class _309_Best_Time_to_Buy_and_Sell_Stock_with_Cooldown {
// ver 1.0
//	public int maxProfit(int[] prices) {
//		if(prices == null || prices.length == 0)
//      	return 0;
//		int n = prices.length;
//		int[] B = new int[n];
//		int[] S = new int[n];
//		int[] C = new int[n];
//		B[0] = -prices[0];
//		
//		for(int i = 1; i < n; i++){
//			C[i] = Math.max(S[i-1], C[i-1]);
//			B[i] = Math.max(C[i-1] - prices[i], B[i-1]);
//			S[i] = B[i-1] + prices[i];
//		}
//		return Math.max(S[n-1], C[n-1]);
//	}
	
	public int maxProfit(int[] prices) {
		if(prices == null || prices.length == 0)
            return 0;
		
		int n = prices.length;
		int B = -prices[0], S=0, C=0;
		
		for(int i = 1; i < n; i++){
			int C1 = Math.max(S, C);
			int B1 = Math.max(C - prices[i], B);
			int S1 = B + prices[i];
			C = C1;
			B = B1;
			S = S1;
		}
		return Math.max(S, C);
	}

	public static void main(String[] args) {
		System.out.println(new _309_Best_Time_to_Buy_and_Sell_Stock_with_Cooldown().maxProfit(new int[]{1,2,3,0,9,2,4,5,1,4,2}));
	}

}
